libnetwork/network/interface.go

```diff
@@ -52,146 +47,94 @@ func NetworkBackend(store storage.Store, conf *config.Config, syslog bool) (type
+	})
+	return netInt, err
 }
 
 func defaultNetworkBackend(store storage.Store, conf *config.Config) (backend types.NetworkBackend, err error) {
-	// read defaultNetworkBackend file
+	err = nil
+
 	file := filepath.Join(store.GraphRoot(), defaultNetworkBackendFileName)
-	b, err := os.ReadFile(file)
-	if err == nil {
-		val := string(b)
-		if val == string(types.Netavark) {
-			return types.Netavark, nil
-		}
-		if val == string(types.CNI) {
-			return types.CNI, nil
-		}
-		return "", fmt.Errorf("unknown network backend value %q in %q", val, file)
-	}
-	// fail for all errors except ENOENT
-	if !errors.Is(err, os.ErrNotExist) {
-		return "", fmt.Errorf("could not read network backend value: %w", err)
-	}
 
-	// cache the network backend to make sure always the same one will be used
-	defer func() {
+	writeBackendToFile := func(backendT types.NetworkBackend) {
 		// only write when there is no error
 		if err == nil {
-			if err := ioutils.AtomicWriteFile(file, []byte(backend), 0o644); err != nil {
+			if err := ioutils.AtomicWriteFile(file, []byte(backendT), 0o644); err != nil {
 				logrus.Errorf("could not write network backend to file: %v", err)
 			}
 		}
-	}()
-
-	_, err = conf.FindHelperBinary("netavark", false)
-	if err != nil {
-		// if we cannot find netavark use CNI
-		return types.CNI, nil
-	}
-
-	// If there are any containers then return CNI
-	cons, err := store.Containers()
-	if err != nil {
-		return "", err
-	}
-	if len(cons) != 0 {
-		return types.CNI, nil
-	}
-
-	// If there are any non ReadOnly images then return CNI
-	imgs, err := store.Images()
-	if err != nil {
-		return "", err
-	}
-	for _, i := range imgs {
-		if !i.ReadOnly {
-			return types.CNI, nil
-		}
 	}
 
-	// If there are CNI Networks then return CNI
-	cniInterface, err := getCniInterface(conf)
+	// read defaultNetworkBackend file
+	b, err := os.ReadFile(file)
 	if err == nil {
-		nets, err := cniInterface.NetworkList()
-		// there is always a default network so check > 1
-		if err != nil && !errors.Is(err, os.ErrNotExist) {
-			return "", err
-		}
+		val := string(b)
 
-		if len(nets) > 1 {
-			// we do not have a fresh system so use CNI
-			return types.CNI, nil
+		// if the network backend has been already set previously,
+		// handle the values depending on whether CNI is supported and
+		// whether the network backend is explicitly configured
+		if val == string(types.Netavark) {
+			// netavark is always good
+			return types.Netavark, nil
+		} else if val == string(types.CNI) {
+			if cniSupported {
+				return types.CNI, nil
+			}
+			// the user has *not* configured a network
+			// backend explicitly but used CNI in the past
+			// => we upgrade them in this case to netavark only
+			writeBackendToFile(types.Netavark)
+			logrus.Info("Migrating network backend to netavark as no backend has been configured previously")
+			return types.Netavark, nil
 		}
+		return "", fmt.Errorf("unknown network backend value %q in %q", val, file)
```

Comment:
I think this line is tripping our [podman tests](https://cirrus-ci.com/task/4772074758078464?logs=main#L1575).

Reply:
To help in debugging: the failure is in f39 rootless and only in two of the e2e tests: [add ssh socket path](https://api.cirrus-ci.com/v1/artifact/task/4772074758078464/html/int-podman-fedora-39-rootless-host-sqlite.log.html#t--podman-system-connection-sshd-and-API-services-required-add-ssh-socket-path-using-connection-heuristic--1) and [kube play using a user namespace](https://api.cirrus-ci.com/v1/artifact/task/4772074758078464/html/int-podman-fedora-39-rootless-host-sqlite.log.html#t--Podman-kube-play-using-a-user-namespace--1), and both failures look like this:
```
→ Enter [It] (test name)
Error: Error: failed to get default network backend: unknown network backend value "" in "[share]/containers/storage/defaultNetworkBackend"
...
No future change is possible.  Bailing out early after 0.741s.
```
I have no idea what "No future change is possible" means, it seems to be coming from ginkgo itself. All other e2e tests pass (1985 of them). I have no idea what is special about those two tests. I'm just posting direct links in hopes that someone else can understand the failure.

Reply:
@edsantiago The test run that @ashley-cui linked includes a lot more test failures. Could it be that there is some leftover polluted cache maybe?

Reply:
The fact that it says `unknown network backend value ""` seems to suggest there is a conditioning in which it writes (or reads) an empty file which seems very bad.

Reply:
@dcermak  @edsantiago  FWIW I changed [the test pr](https://github.com/containers/podman/pull/21179/checks?check_run_id=20741136641) to always build CNI, and it looks like all the rootless runs are failing. I think this file should only be created by this network backend code? I can try to help dig a little failure but @Luap99 's  diagnosis seems correct.

Reply:
@ashley-cui could you please rebase the test PRs? I've fixed the issue that @Luap99 found

Reply:
Rebased!